poco: 1.13.3 -> 1.14.1

trofi opened the pull request:
Changes:
- https://github.com/pocoproject/poco/releases/tag/poco-1.14.0-release
- https://github.com/pocoproject/poco/releases/tag/poco-1.14.1-release


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

trofi committed:
poco: 1.13.3 -> 1.14.1

Changes:
- https://github.com/pocoproject/poco/releases/tag/poco-1.14.0-release
- https://github.com/pocoproject/poco/releases/tag/poco-1.14.1-release

trofi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 369312`

---
### `x86_64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>collabora-online</li>
    <li>craftos-pc</li>
    <li>pothos</li>
    <li>vscode-extensions.jackmacwindows.craftos-pc</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 11 packages built:</summary>
  <ul>
    <li>ioq3-scion</li>
    <li>ioquake3</li>
    <li>litebrowser</li>
    <li>multipass</li>
    <li>mumble</li>
    <li>mumble_overlay</li>
    <li>murmur</li>
    <li>poco</li>
    <li>poco.dev</li>
    <li>quake3demo</li>
    <li>sanjuuni</li>
  </ul>
</details>

trofi commented:
All failures are regressions. Converting to draft.

dwt commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 369312`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>litebrowser</li>
    <li>sanjuuni</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>poco</li>
    <li>poco.dev</li>
  </ul>
</details>

dwt commented:
This  update actually fixes compilation of poco on darwin, as it includes [this fix](https://github.com/cdigruttola/poco/commit/3b79a51722cc2b6bd9484e0c004b17161e61af96), which previously prevented compilation with clang 19. Since I would like to port mumble and ioquake3 which depend on this, is there something I can do to get this moving forward?

I noticed that there is actually already a 1.14.1 release of poco, maybe updating to that could help?

trofi commented:
I updated to `1.14.1`, but I suspect that the API change did not go away and most dependencies still break with new release. Ran the `nixpkgs-review 369312` locally.

dwt commented:
I tried doing this update and running nixpkgs-review locally, but I never got it to build just poco and its dependencies, could you perhaps share your workflow for running nixpkgs-review locally? Perhaps I'm missing something obvious.

I've tried `nixpkgs-review rev HEAD` (that built way too much) and `nixpkgs-review rev HEAD --package poco` (that only built poco, no dependent packages). Is there a different way to test local changes without creating a pr first that focuses better on just the packages changed by a pull request?

dwt commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 369312`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>litebrowser</li>
    <li>sanjuuni</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>poco</li>
    <li>poco.dev</li>
  </ul>
</details>

dwt commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 369312`

---
### `aarch64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>collabora-online</li>
    <li>craftos-pc</li>
    <li>pothos</li>
    <li>vscode-extensions.jackmacwindows.craftos-pc</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>ioq3-scion</li>
    <li>ioquake3</li>
    <li>litebrowser</li>
    <li>mumble</li>
    <li>murmur</li>
    <li>poco</li>
    <li>poco.dev</li>
    <li>quake3demo</li>
    <li>sanjuuni</li>
  </ul>
</details>

trofi commented:
> I tried doing this update and running nixpkgs-review locally, but I never got it to build just poco and its dependencies, could you perhaps share your workflow for running nixpkgs-review locally? Perhaps I'm missing something obvious.
> 
> I've tried `nixpkgs-review rev HEAD` (that built way too much) and `nixpkgs-review rev HEAD --package poco` (that only built poco, no dependent packages). Is there a different way to test local changes without creating a pr first that focuses better on just the packages changed by a pull request?

I just run `$ nixpkgs-review pr --no-shell --post-result --build-args '--max-jobs 1' 369312` and wait A Lot (I use a tiny wrapper around it to call it as `nr $pr`: https://github.com/trofi/home/blob/master/bin/nr).

trofi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 369312`

---
### `x86_64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>collabora-online</li>
    <li>craftos-pc</li>
    <li>pothos</li>
    <li>vscode-extensions.jackmacwindows.craftos-pc</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 11 packages built:</summary>
  <ul>
    <li>ioq3-scion</li>
    <li>ioquake3</li>
    <li>litebrowser</li>
    <li>multipass</li>
    <li>mumble</li>
    <li>mumble_overlay</li>
    <li>murmur</li>
    <li>poco</li>
    <li>poco.dev</li>
    <li>quake3demo</li>
    <li>sanjuuni</li>
  </ul>
</details>

dwt commented:
I've looked into the build failures from pothos, and found two problems:
* I'ts not using c++17 to build, which fails because poco 1.14 requires it. Enabling that gets the build further, until it fails again because [this 5 (!) year old upstream fix](https://github.com/pothosware/PothosBlocks/commit/6ea83d86ca6ad8be83bab4f2d46afeb68690d652#diff-c83b83e00ac990aae05b79ab07233780e4827130fcc357266d46edfb504f317f) does not seem to be part of the release we use

That is kind of a shame, as this concrete error is probably easy to fix, and upstream already has some fixes. However, the upstream project looks quite dormant, so I don't have high hopes for an upstream fix. [Even though, I've opened a bug there about the issue](https://github.com/pothosware/PothosBlocks/issues/80)

dwt commented:
I actually did get pothos to compile with this patch:

```diff
diff --git a/pkgs/applications/radio/pothos/cstring.patch b/pkgs/applications/radio/pothos/cstring.patch
new file mode 100644
index 000000000000..4e8957c3f122
--- /dev/null
+++ b/pkgs/applications/radio/pothos/cstring.patch
@@ -0,0 +1,50 @@
+Submodule blocks contains modified content
+diff --git a/blocks/file/BinaryFileSink.cpp b/blocks/file/BinaryFileSink.cpp
+index 31c9a41..0083b0d 100644
+--- a/blocks/file/BinaryFileSink.cpp
++++ b/blocks/file/BinaryFileSink.cpp
+@@ -13,6 +13,7 @@
+ #endif //_MSC_VER
+ #include <stdio.h>
+ #include <cerrno>
++#include <cstring>
+
+ #ifndef O_BINARY
+ #define O_BINARY 0
+diff --git a/blocks/file/BinaryFileSource.cpp b/blocks/file/BinaryFileSource.cpp
+index 0151231..379d383 100644
+--- a/blocks/file/BinaryFileSource.cpp
++++ b/blocks/file/BinaryFileSource.cpp
+@@ -13,6 +13,7 @@
+ #endif //_MSC_VER
+ #include <stdio.h>
+ #include <cerrno>
++#include <cstring>
+
+ #ifndef O_BINARY
+ #define O_BINARY 0
+diff --git a/blocks/file/TextFileSink.cpp b/blocks/file/TextFileSink.cpp
+index b4b2f08..2be66e2 100644
+--- a/blocks/file/TextFileSink.cpp
++++ b/blocks/file/TextFileSink.cpp
+@@ -6,6 +6,7 @@
+ #include <complex>
+ #include <fstream>
+ #include <cerrno>
++#include <cstring>
+
+ /***********************************************************************
+  * |PothosDoc Text File Sink
+Submodule soapy contains modified content
+diff --git a/soapy/DemoController.cpp b/soapy/DemoController.cpp
+index 4ce8ead..9a4e742 100644
+--- a/soapy/DemoController.cpp
++++ b/soapy/DemoController.cpp
+@@ -6,6 +6,7 @@
+ #include <iostream>
+ #include <chrono>
+ #include <algorithm> //min/max
++#include <cstring>
+
+ /***********************************************************************
+  * |PothosDoc SDR Demo Controller
diff --git a/pkgs/applications/radio/pothos/default.nix b/pkgs/applications/radio/pothos/default.nix
index 641df521c7fd..7d8ac58cf6ee 100644
--- a/pkgs/applications/radio/pothos/default.nix
+++ b/pkgs/applications/radio/pothos/default.nix
@@ -41,6 +41,8 @@ mkDerivation rec {
       url = "https://github.com/pothosware/PothosCore/commit/092d1209b0fd0aa8a1733706c994fa95e66fd017.patch";
       hash = "sha256-bZXG8kD4+1LgDV8viZrJ/DMjg8UvW7b5keJQDXurfkA=";
     })
+    # various source files of pathos are missing imports of <cstring>
+    ./cstring.patch
   ];
 
   nativeBuildInputs = [
@@ -64,6 +66,8 @@ mkDerivation rec {
     python3
   ];
 
+  NIX_CFLAGS_COMPILE = [ "-std=c++17" ];
+
   postInstall = ''
     install -Dm644 $out/share/Pothos/Desktop/pothos-flow.desktop $out/share/applications/pothos-flow.desktop
     install -Dm644 $out/share/Pothos/Desktop/pothos-flow-16.png $out/share/icons/hicolor/16x16/apps/pothos-flow.png
```

Caveat: Even though I have a linux builder, I don't (yet) have a linux machine capable of actually testing this package.

@trofi Do you think it viable to include something like that in in this pull request? Do you think fixing the broken packages in this way a viable way forward to get this poco update into nixpkgs?

trofi commented:
Yeah, that sounds great. If you want you can sent PRs against https://github.com/trofi/nixpkgs/tree/poco-update and I'll apply them to appear in this proposal.

trofi commented:
Or alternatively I can close this PR so you could create the PR and fix all the fallouts.

dwt commented:
I'd be happy to send you pull requests against your branch.

dwt committed:
pothos: fix the build against poco-1.14

dwt commented:
[Here you go](https://github.com/trofi/nixpkgs/pull/2)

dwt commented:
> I just run `$ nixpkgs-review pr --no-shell --post-result --build-args '--max-jobs 1' 369312` and wait A Lot (I use a tiny wrapper around it to call it as `nr $pr`: https://github.com/trofi/home/blob/master/bin/nr).

That means you create the pull request first and only then test using `nixpkgs-review`? I was hoping for a workflow where I could check before posting the pull request.

trofi commented:
> [Here you go](https://github.com/trofi/nixpkgs/pull/2)

Pulled. Thank you!

dwt commented:
Investigating the build further, I found that the option

```nix
 configureFlags = [
   "--unbundled"
 ];
```

does not actually have any effect, as cmake is used to build this. [It was changed from a cmake flag to a configure flag a while back](https://github.com/NixOS/nixpkgs/commit/369650ef388e4a6488b1783dc747cab7559a90d4), to fix the `nix build .#pkgsStatic.poco` build. That earlier fix seems faulty - not sure it was ever correct.  Reverting that, I found that there was actually a dependency missing, and upstream seems to have [a bug with static builds](https://github.com/pocoproject/poco/issues/4871). [I have provided a workaround as a second pull request to you @trofi](https://github.com/trofi/nixpkgs/pull/3).

trofi commented:
Pulled the PR and slightly reworded the comment

dwt commented:
> Pulled the PR and slightly reworded the comment

Sorry, I did push some additional commits that also fixed the comments. For example, the patch actually disables the internal  pcre artifacts for _static_ builds, not all the time. :/

dwt committed:
craftos-pc: fix the build against poco-1.14

Upstream seems to have errors, where it doesn't
include all the poco headers it requires for it
to build. I've provided a patch and
[opened a pull request upstream to get it merged](https://github.com/MCJack123/craftos2/issues/391].

dwt commented:
[And another pull request to get craftos-pc compiling with this poco version](https://github.com/trofi/nixpkgs/pull/5)

dwt committed:
collabora-online: fix build against poco-1.14

This adds some missing includes for collabora online to actually use
the poco-json library.

Upstream pull request at
https://github.com/CollaboraOnline/online/issues/11175

dwt commented:
[And the last pull request to fix the build of collabora-online](https://github.com/trofi/nixpkgs/pull/6)

dwt commented:
I think this should be it, as far as I can see this fixes all the builds that are failing because of the poco update. Do you see anything else that is missing to make this mergeable?

trofi commented:
> [And the last pull request to fix the build of collabora-online](https://github.com/trofi/nixpkgs/pull/6)

Pulled it, rebased, squashed and renamed commits to have consistent `$pkg:` prefix.

dwt commented:
Ah damn. More work.

dwt commented:
[And another one for mumble](https://github.com/trofi/nixpkgs/pull/8)

As I was not yet able to find good documentation: At what time would I consider using a `propagatedBuildInput` for something like `utf8proc`? If all the libraries that use `poco` need it, should it then become a propagated dependency?

dwt commented:
Not quite sure what's up with `multipass` my build fails with an error like this, that seems... odd?

```
┃ error: builder for '/nix/store/76m5rq1q63h3pm2c4ql1myx673c06zc8-flutter-cache-dir.drv' failed with exit code 1;
┃        last 1 log lines:
┃        > error: executing '/nix/store/ck4s3x3jvml21zpny04xg7fbi8qpxjax-bash-5.2p37/bin/bash': Undefined error: 0
┃        For full logs, run 'nix log /nix/store/76m5rq1q63h3pm2c4ql1myx673c06zc8-flutter-cache-dir.drv'.
```

@trofi what kind of failures do you get?

trofi commented:
```
$ nix build --no-link github:NixOS/nixpkgs/pull/369312/merge#multipass -L
...
multipassd> CMake Error at /nix/store/wjb1rq9xgxzq6n1b7pbrbsk5lrdn9qc7-cmake-3.31.3/share/cmake-3.31/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
multipassd>   Could NOT find Utf8Proc (missing: UTF8PROC_LIBRARY UTF8PROC_INCLUDE_DIR)
```

trofi commented:
Let me close this PR. I don't provide much value here. Please reopen a new one with your fixes. I can run build tests against it and provide some basic explanation around build failures.

dwt commented:
Well, the nix-know-how you provide is (and will be) much appreciated. Since I'm very much new to nix (just started using it a few weeks ago) I very much don't know how to do to many things and wether they are acceptable or right / idiomatic for nix. So hopefully I can count on your know how for those things in that follow-up pull request.

dwt commented:
For all upstream projects that where linked to this pull request: please see the continuation of this work at #383163
